File: src/calibration/checkerboard.cpp
#include "chess_robot/calibration/checkerboard.hpp"

#include "chess_robot/utils/chessboard.hpp"

#include <opencv2/calib3d.hpp>

namespace chess_robot {

std::optional<cv::Mat> findCheckerboardCorners(const cv::Mat& frame,
                                               cv::Size pattern_size,
                                               bool robust) {
  return findInnerCorners(frame, pattern_size, robust);
}

cv::Mat drawCheckerboardOverlay(const cv::Mat& frame,
                                cv::Size pattern_size,
                                const cv::Mat& corners) {
  cv::Mat overlay = frame.clone();
  cv::drawChessboardCorners(overlay, pattern_size, corners, true);
  return overlay;
}

cv::Mat buildObjectPoints(int cols, int rows, float square_size_mm) {
  cv::Mat grid(rows * cols, 1, CV_32FC3);
  int index = 0;
  for (int row = 0; row < rows; ++row) {
    for (int col = 0; col < cols; ++col) {
      grid.at<cv::Vec3f>(index++) = cv::Vec3f(col * square_size_mm, row * square_size_mm, 0.0f);
    }
  }
  return grid;
}

}  // namespace chess_robot

File: python/src/calibration/checkerboard.py
"""Checkerboard detection for intrinsic calibration."""

from __future__ import annotations

import cv2
import numpy as np

from src.utils.chessboard import find_inner_corners


def find_checkerboard_corners(
    frame: np.ndarray,
    pattern_size: tuple[int, int],
    *,
    robust: bool = True,
) -> tuple[bool, np.ndarray | None]:
    corners = find_inner_corners(frame, pattern_size, robust=robust)
    if corners is None:
        return False, None
    return True, corners


def draw_checkerboard_overlay(
    frame: np.ndarray,
    pattern_size: tuple[int, int],
    corners: np.ndarray,
) -> np.ndarray:
    overlay = frame.copy()
    cv2.drawChessboardCorners(overlay, pattern_size, corners, True)
    return overlay


def build_object_points(
    cols: int,
    rows: int,
    square_size_mm: float,
) -> np.ndarray:
    grid = np.zeros((rows * cols, 3), np.float32)
    grid[:, :2] = np.mgrid[0:cols, 0:rows].T.reshape(-1, 2)
    grid *= square_size_mm
    return grid
